File: web/src/lib/crypto/vectors.test.ts
// Verifies this package against the SAME fixed vectors the Go
// implementation checks itself against (internal/crypto/testdata/vectors.json,
// #20). This is the #24 half that didn't exist yet: proving the TypeScript
// side agrees with Go's, not just with its own round-trips.
//
// Reads the vectors file directly from the Go package rather than a copy —
// there must be exactly one vectors.json in the repo, or the two
// implementations could silently drift against different "shared" files.

import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'

import { decrypt, DecryptionFailedError, encryptWithNonce } from './aesgcm.js'
import { deriveKey } from './argon2.js'
import * as ed25519 from './ed25519.js'
import { fingerprint } from './fingerprint.js'
import { bytesToHex, hexToBytes } from './hex.js'
import { signedPayload } from './payload.js'
import { unwrap, wrapWithEphemeralAndNonce } from './x25519.js'

const VECTORS_PATH = fileURLToPath(
  new URL('../../../../internal/crypto/testdata/vectors.json', import.meta.url),
)

interface VectorFile {
  version: number
  kdf: {
    password: string
    salt_hex: string
    m_kib: number
    t: number
    p: number
    key_hex: string
  }[]
  aead: {
    name: string
    key_hex: string
    nonce_hex: string
    plaintext_hex: string
    aad_hex: string
    ciphertext_hex: string
  }[]
  aead_negative: {
    name: string
    key_hex: string
    nonce_hex: string
    ciphertext_hex: string
    aad_hex: string
    wrong_aad_hex: string
  }[]
  signing: {
    name: string
    private_key_hex: string
    public_key_hex: string
    context: string
    message_hex: string
    signature_hex: string
  }[]
  signed_payload: {
    name: string
    private_key_hex: string
    public_key_hex: string
    context: string
    author_uuid: string
    group_id: string
    sort_key: string
    generation: number
    utc_day: string
    ciphertext_hex: string
    payload_hex: string
    signature_hex: string
  }[]
  wrapping: {
    name: string
    recipient_private_key_hex: string
    recipient_public_key_hex: string
    ephemeral_private_key_hex: string
    ephemeral_public_key_hex: string
    plaintext_hex: string
    aad_hex: string
    wrapped_nonce_hex: string
    wrapped_ciphertext_hex: string
  }[]
  genkey_chain: {
    name: string
    gen_n_key_hex: string
    gen_n_plus_1_key_hex: string
    aad_hex: string
    link_nonce_hex: string
    link_ciphertext_hex: string
    forward_must_fail: boolean
  }[]
  fingerprint: {
    name: string
    signing_public_key_hex: string
    wrapping_public_key_hex: string
    fingerprint: string
  }[]
}

const vectors: VectorFile = JSON.parse(readFileSync(VECTORS_PATH, 'utf8'))

describe('kdf vectors', () => {
  for (const tc of vectors.kdf) {
    it(`m=${tc.m_kib}KiB t=${tc.t} p=${tc.p}`, async () => {
      const salt = hexToBytes(tc.salt_hex)
      const want = hexToBytes(tc.key_hex)
      const got = await deriveKey(
        tc.password,
        salt,
        { memoryKiB: tc.m_kib, iterations: tc.t, parallelism: tc.p },
        want.length,
      )
      expect(bytesToHex(got)).toBe(tc.key_hex)
    })
  }
})

describe('aead vectors', () => {
  for (const tc of vectors.aead) {
    it(tc.name, async () => {
      const key = hexToBytes(tc.key_hex)
      const nonce = hexToBytes(tc.nonce_hex)
      const plaintext = hexToBytes(tc.plaintext_hex)
      const aad = hexToBytes(tc.aad_hex)

      const ciphertext = await encryptWithNonce(key, nonce, plaintext, aad)
      expect(bytesToHex(ciphertext)).toBe(tc.ciphertext_hex)

      const decrypted = await decrypt(key, nonce, ciphertext, aad)
      expect(bytesToHex(decrypted)).toBe(tc.plaintext_hex)
    })
  }
})

describe('aead negative vectors', () => {
  for (const tc of vectors.aead_negative) {
    it(tc.name, async () => {
      const key = hexToBytes(tc.key_hex)
      const nonce = hexToBytes(tc.nonce_hex)
      const ciphertext = hexToBytes(tc.ciphertext_hex)
      const wrongAad = hexToBytes(tc.wrong_aad_hex)

      await expect(decrypt(key, nonce, ciphertext, wrongAad)).rejects.toThrow(DecryptionFailedError)
    })
  }
})

describe('signing vectors', () => {
  for (const tc of vectors.signing) {
    it(tc.name, () => {
      const key = ed25519.fromGoPrivateKeyBytes(hexToBytes(tc.private_key_hex))
      expect(bytesToHex(key.publicKey)).toBe(tc.public_key_hex)

      const message = hexToBytes(tc.message_hex)
      const signature = ed25519.sign(key, tc.context as ed25519.SigningContext, message)
      expect(bytesToHex(signature)).toBe(tc.signature_hex)

      expect(
        ed25519.verify(key.publicKey, tc.context as ed25519.SigningContext, message, signature),
      ).toBe(true)
    })
  }
})

describe('signed payload vectors', () => {
  for (const tc of vectors.signed_payload) {
    it(tc.name, () => {
      const key = ed25519.fromGoPrivateKeyBytes(hexToBytes(tc.private_key_hex))
      const ciphertext = hexToBytes(tc.ciphertext_hex)

      const payload = signedPayload(
        tc.author_uuid,
        tc.group_id,
        tc.sort_key,
        tc.generation,
        tc.utc_day,
        ciphertext,
      )
      expect(bytesToHex(payload)).toBe(tc.payload_hex)

      const signature = ed25519.sign(key, tc.context as ed25519.SigningContext, payload)
      expect(bytesToHex(signature)).toBe(tc.signature_hex)
    })
  }
})

describe('wrapping vectors', () => {
  for (const tc of vectors.wrapping) {
    it(tc.name, async () => {
      const recipientPriv = hexToBytes(tc.recipient_private_key_hex)
      const recipientPub = hexToBytes(tc.recipient_public_key_hex)
      const ephemeralPriv = hexToBytes(tc.ephemeral_private_key_hex)
      const plaintext = hexToBytes(tc.plaintext_hex)
      const aad = hexToBytes(tc.aad_hex)
      const nonce = hexToBytes(tc.wrapped_nonce_hex)

      const wrapped = await wrapWithEphemeralAndNonce(
        recipientPub,
        ephemeralPriv,
        nonce,
        plaintext,
        aad,
      )
      expect(bytesToHex(wrapped.ephemeralPub)).toBe(tc.ephemeral_public_key_hex)
      expect(bytesToHex(wrapped.ciphertext)).toBe(tc.wrapped_ciphertext_hex)

      const unwrapped = await unwrap(recipientPriv, wrapped, aad)
      expect(bytesToHex(unwrapped)).toBe(tc.plaintext_hex)
    })
  }
})

// The negative vector issue #20 (and #24 now) calls out by name: GENKEY#<n>
// holds generation n encrypted under generation n+1's key. A holder of
// generation n+1 decrypts backward to generation n; a holder of only
// generation n has no key material that decrypts a link encrypted under
// generation n+1. This exercises the plain AES-GCM primitives, not X25519 —
// see internal/crypto/vectors_test.go's TestVectorGenkeyChain for why using
// two unrelated asymmetric keys here would test the wrong construction.
describe('genkey chain vectors', () => {
  for (const tc of vectors.genkey_chain) {
    it(tc.name, async () => {
      const genN = hexToBytes(tc.gen_n_key_hex)
      const genNPlus1 = hexToBytes(tc.gen_n_plus_1_key_hex)
      const aad = hexToBytes(tc.aad_hex)
      const nonce = hexToBytes(tc.link_nonce_hex)

      const ciphertext = await encryptWithNonce(genNPlus1, nonce, genN, aad)
      expect(bytesToHex(ciphertext)).toBe(tc.link_ciphertext_hex)

      const decrypted = await decrypt(genNPlus1, nonce, ciphertext, aad)
      expect(bytesToHex(decrypted)).toBe(tc.gen_n_key_hex)

      if (tc.forward_must_fail) {
        await expect(decrypt(genN, nonce, ciphertext, aad)).rejects.toThrow(DecryptionFailedError)
      }
    })
  }
})

describe('fingerprint vectors', () => {
  for (const tc of vectors.fingerprint) {
    it(tc.name, () => {
      const signingPub = hexToBytes(tc.signing_public_key_hex)
      const wrappingPub = hexToBytes(tc.wrapping_public_key_hex)
      expect(fingerprint(signingPub, wrappingPub)).toBe(tc.fingerprint)
    })
  }
})

File: internal/crypto/vectors_test.go
package crypto

import (
	"bytes"
	"crypto/ecdh"
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"

	"golang.org/x/crypto/argon2"
)

// vectorFile mirrors internal/crypto/testdata/vectors.json. Field shapes
// here must match internal/crypto/testdata/gen/main.go's output types
// exactly, since both are the same JSON schema — see testdata/README.md.
type vectorFile struct {
	Version       int                   `json:"version"`
	KDF           []kdfVector           `json:"kdf"`
	AEAD          []aeadVector          `json:"aead"`
	AEADNegative  []aeadNegativeVector  `json:"aead_negative"`
	Signing       []signingVector       `json:"signing"`
	SignedPayload []signedPayloadVector `json:"signed_payload"`
	Wrapping      []wrapVector          `json:"wrapping"`
	GenkeyChain   []genkeyChainVector   `json:"genkey_chain"`
	Fingerprint   []fingerprintVector   `json:"fingerprint"`
}

type kdfVector struct {
	Password string `json:"password"`
	SaltHex  string `json:"salt_hex"`
	M        uint32 `json:"m_kib"`
	T        uint32 `json:"t"`
	P        uint8  `json:"p"`
	KeyHex   string `json:"key_hex"`
}

type aeadVector struct {
	Name          string `json:"name"`
	KeyHex        string `json:"key_hex"`
	NonceHex      string `json:"nonce_hex"`
	PlaintextHex  string `json:"plaintext_hex"`
	AADHex        string `json:"aad_hex"`
	CiphertextHex string `json:"ciphertext_hex"`
}

type aeadNegativeVector struct {
	Name          string `json:"name"`
	KeyHex        string `json:"key_hex"`
	NonceHex      string `json:"nonce_hex"`
	CiphertextHex string `json:"ciphertext_hex"`
	AADHex        string `json:"aad_hex"`
	WrongAADHex   string `json:"wrong_aad_hex"`
}

type signingVector struct {
	Name         string `json:"name"`
	PrivateHex   string `json:"private_key_hex"`
	PublicHex    string `json:"public_key_hex"`
	Context      string `json:"context"`
	MessageHex   string `json:"message_hex"`
	SignatureHex string `json:"signature_hex"`
}

type signedPayloadVector struct {
	Name          string `json:"name"`
	PrivateHex    string `json:"private_key_hex"`
	PublicHex     string `json:"public_key_hex"`
	Context       string `json:"context"`
	AuthorUUID    string `json:"author_uuid"`
	GroupID       string `json:"group_id"`
	SortKey       string `json:"sort_key"`
	Generation    uint64 `json:"generation"`
	UTCDay        string `json:"utc_day"`
	CiphertextHex string `json:"ciphertext_hex"`
	PayloadHex    string `json:"payload_hex"`
	SignatureHex  string `json:"signature_hex"`
}

type wrapVector struct {
	Name              string `json:"name"`
	RecipientPrivHex  string `json:"recipient_private_key_hex"`
	RecipientPubHex   string `json:"recipient_public_key_hex"`
	EphemeralPrivHex  string `json:"ephemeral_private_key_hex"`
	EphemeralPubHex   string `json:"ephemeral_public_key_hex"`
	PlaintextHex      string `json:"plaintext_hex"`
	AADHex            string `json:"aad_hex"`
	WrappedNonceHex   string `json:"wrapped_nonce_hex"`
	WrappedCiphertext string `json:"wrapped_ciphertext_hex"`
}

type genkeyChainVector struct {
	Name            string `json:"name"`
	GenNKeyHex      string `json:"gen_n_key_hex"`
	GenNPlus1KeyHex string `json:"gen_n_plus_1_key_hex"`
	AADHex          string `json:"aad_hex"`
	LinkNonceHex    string `json:"link_nonce_hex"`
	LinkCiphertext  string `json:"link_ciphertext_hex"`
	ForwardMustFail bool   `json:"forward_must_fail"`
}

type fingerprintVector struct {
	Name           string `json:"name"`
	SigningPubHex  string `json:"signing_public_key_hex"`
	WrappingPubHex string `json:"wrapping_public_key_hex"`
	Fingerprint    string `json:"fingerprint"`
}

func loadVectors(t *testing.T) vectorFile {
	t.Helper()
	data, err := os.ReadFile("testdata/vectors.json")
	if err != nil {
		t.Fatalf("reading testdata/vectors.json: %v", err)
	}
	var v vectorFile
	if err := json.Unmarshal(data, &v); err != nil {
		t.Fatalf("parsing testdata/vectors.json: %v", err)
	}
	return v
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("invalid hex %q: %v", s, err)
	}
	return b
}

func TestVectorKDF(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.KDF {
		t.Run(tc.Password, func(t *testing.T) {
			salt := mustHex(t, tc.SaltHex)
			want := mustHex(t, tc.KeyHex)
			// Parameters are read from the vector itself, not a compiled-in
			// constant, per docs/DESIGN.md's "Testing" section — this fails
			// if either the numbers or the read path changes.
			got := argon2.IDKey([]byte(tc.Password), salt, tc.T, tc.M, tc.P, uint32(len(want)))
			if !bytes.Equal(got, want) {
				t.Fatalf("Argon2id(m=%d,t=%d,p=%d) = %x, want %x", tc.M, tc.T, tc.P, got, want)
			}
		})
	}
}

func TestVectorAEAD(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.AEAD {
		t.Run(tc.Name, func(t *testing.T) {
			key := mustHex(t, tc.KeyHex)
			nonce := mustHex(t, tc.NonceHex)
			plaintext := mustHex(t, tc.PlaintextHex)
			aad := mustHex(t, tc.AADHex)
			want := mustHex(t, tc.CiphertextHex)

			got, err := EncryptWithNonce(key, nonce, plaintext, aad)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(got, want) {
				t.Fatalf("EncryptWithNonce = %x, want %x", got, want)
			}

			decrypted, err := Decrypt(key, nonce, want, aad)
			if err != nil {
				t.Fatalf("Decrypt: %v", err)
			}
			if !bytes.Equal(decrypted, plaintext) {
				t.Fatalf("Decrypt = %x, want %x", decrypted, plaintext)
			}
		})
	}
}

// TestVectorAEADNegative is the negative case docs/DESIGN.md's "Testing"
// section calls out by name: a mismatched AAD must fail, which is what
// catches an implementation that quietly stops binding ciphertext to its
// address.
func TestVectorAEADNegative(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.AEADNegative {
		t.Run(tc.Name, func(t *testing.T) {
			key := mustHex(t, tc.KeyHex)
			nonce := mustHex(t, tc.NonceHex)
			ciphertext := mustHex(t, tc.CiphertextHex)
			wrongAAD := mustHex(t, tc.WrongAADHex)

			if _, err := Decrypt(key, nonce, ciphertext, wrongAAD); err != ErrDecryptionFailed {
				t.Fatalf("Decrypt with wrong AAD: got err %v, want ErrDecryptionFailed", err)
			}
		})
	}
}

func TestVectorSigning(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.Signing {
		t.Run(tc.Name, func(t *testing.T) {
			priv := ed25519.PrivateKey(mustHex(t, tc.PrivateHex))
			pub := ed25519.PublicKey(mustHex(t, tc.PublicHex))
			message := mustHex(t, tc.MessageHex)
			want := mustHex(t, tc.SignatureHex)

			got, err := Sign(priv, SigningContext(tc.Context), message)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(got, want) {
				t.Fatalf("Sign = %x, want %x", got, want)
			}
			if !Verify(pub, SigningContext(tc.Context), message, want) {
				t.Fatal("Verify rejected the vector's own signature")
			}
		})
	}
}

func TestVectorSignedPayload(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.SignedPayload {
		t.Run(tc.Name, func(t *testing.T) {
			priv := ed25519.PrivateKey(mustHex(t, tc.PrivateHex))
			pub := ed25519.PublicKey(mustHex(t, tc.PublicHex))
			ciphertext := mustHex(t, tc.CiphertextHex)
			wantPayload := mustHex(t, tc.PayloadHex)
			wantSig := mustHex(t, tc.SignatureHex)

			gotPayload := SignedPayload(tc.AuthorUUID, tc.GroupID, tc.SortKey, tc.Generation, tc.UTCDay, ciphertext)
			if !bytes.Equal(gotPayload, wantPayload) {
				t.Fatalf("SignedPayload = %x, want %x", gotPayload, wantPayload)
			}

			gotSig, err := Sign(priv, SigningContext(tc.Context), gotPayload)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(gotSig, wantSig) {
				t.Fatalf("Sign(payload) = %x, want %x", gotSig, wantSig)
			}
			if !Verify(pub, SigningContext(tc.Context), gotPayload, wantSig) {
				t.Fatal("Verify rejected the vector's own signature over SignedPayload")
			}
		})
	}
}

func TestVectorWrapping(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.Wrapping {
		t.Run(tc.Name, func(t *testing.T) {
			recipientPriv, err := ecdh.X25519().NewPrivateKey(mustHex(t, tc.RecipientPrivHex))
			if err != nil {
				t.Fatal(err)
			}
			ephemeralPriv, err := ecdh.X25519().NewPrivateKey(mustHex(t, tc.EphemeralPrivHex))
			if err != nil {
				t.Fatal(err)
			}
			plaintext := mustHex(t, tc.PlaintextHex)
			aad := mustHex(t, tc.AADHex)
			wantNonce := mustHex(t, tc.WrappedNonceHex)
			wantCiphertext := mustHex(t, tc.WrappedCiphertext)

			wrapped, err := WrapWithEphemeralAndNonce(recipientPriv.PublicKey(), ephemeralPriv, wantNonce, plaintext, aad)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(wrapped.Nonce, wantNonce) {
				t.Fatalf("nonce = %x, want %x", wrapped.Nonce, wantNonce)
			}
			if !bytes.Equal(wrapped.Ciphertext, wantCiphertext) {
				t.Fatalf("ciphertext = %x, want %x", wrapped.Ciphertext, wantCiphertext)
			}
			if !bytes.Equal(wrapped.EphemeralPub, ephemeralPriv.PublicKey().Bytes()) {
				t.Fatalf("ephemeral pub = %x, want %x", wrapped.EphemeralPub, ephemeralPriv.PublicKey().Bytes())
			}

			got, err := Unwrap(recipientPriv, wrapped, aad)
			if err != nil {
				t.Fatalf("Unwrap: %v", err)
			}
			if !bytes.Equal(got, plaintext) {
				t.Fatalf("Unwrap = %x, want %x", got, plaintext)
			}
		})
	}
}

// TestVectorGenkeyChain is the negative vector issue #20 calls out by name:
// GENKEY#<n> holds generation n (a symmetric AES-256 key) encrypted under
// generation n+1's key, per docs/DESIGN.md:1131. A holder of generation n+1
// decrypts backward to generation n; a holder of only generation n has no
// key material that decrypts anything encrypted under generation n+1.
func TestVectorGenkeyChain(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.GenkeyChain {
		t.Run(tc.Name, func(t *testing.T) {
			genN := mustHex(t, tc.GenNKeyHex)
			genNPlus1 := mustHex(t, tc.GenNPlus1KeyHex)
			aad := mustHex(t, tc.AADHex)
			nonce := mustHex(t, tc.LinkNonceHex)
			wantCiphertext := mustHex(t, tc.LinkCiphertext)

			gotCiphertext, err := EncryptWithNonce(genNPlus1, nonce, genN, aad)
			if err != nil {
				t.Fatal(err)
			}
			if !bytes.Equal(gotCiphertext, wantCiphertext) {
				t.Fatalf("link ciphertext = %x, want %x", gotCiphertext, wantCiphertext)
			}

			got, err := Decrypt(genNPlus1, nonce, wantCiphertext, aad)
			if err != nil {
				t.Fatalf("gen N+1 holder failed to decrypt the link: %v", err)
			}
			if !bytes.Equal(got, genN) {
				t.Fatalf("gen N+1 holder decrypted %x, want %x", got, genN)
			}

			if tc.ForwardMustFail {
				if _, err := Decrypt(genN, nonce, wantCiphertext, aad); err != ErrDecryptionFailed {
					t.Fatalf("gen N holder decrypted a link encrypted under gen N+1: got err %v, want ErrDecryptionFailed", err)
				}
			}
		})
	}
}

func TestVectorFingerprint(t *testing.T) {
	v := loadVectors(t)
	for _, tc := range v.Fingerprint {
		t.Run(tc.Name, func(t *testing.T) {
			signingPub := ed25519.PublicKey(mustHex(t, tc.SigningPubHex))
			wrappingPub := mustHex(t, tc.WrappingPubHex)

			got, err := Fingerprint(signingPub, wrappingPub)
			if err != nil {
				t.Fatal(err)
			}
			if got != tc.Fingerprint {
				t.Fatalf("Fingerprint = %q, want %q", got, tc.Fingerprint)
			}
		})
	}
}
